recipes-devtools/maccalc/maccalc/main.c

```diff
@@ -0,0 +1,256 @@
+/*
+ * MAC address manupulation utility
+ *
+ * Copyright (C) 2011 Gabor Juhos <juhosg@openwrt.org>
+ *
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 as published
+ * by the Free Software Foundation.
+ *
+ */
+
+#include <errno.h>
+#include <stdlib.h>
+#include <stdio.h>
+#include <stdint.h>
+#include <string.h>
+#include <unistd.h>
+
+#define MAC_ADDRESS_LEN		6
+
+#define ERR_INVALID		1
+#define ERR_IO			2
+
+static void usage(void);
+
+char *maccalc_name;
+
+static int parse_mac(const char *mac_str, unsigned char *buf)
+{
+	int t;
+
+	t = sscanf(mac_str, "%02hhx:%02hhx:%02hhx:%02hhx:%02hhx:%02hhx",
+		   &buf[0], &buf[1], &buf[2], &buf[3], &buf[4], &buf[5]);
+
+	if (t != MAC_ADDRESS_LEN)
+		return ERR_INVALID;
+
+	return 0;
+}
+
+static void print_mac(unsigned char *buf)
+{
+	printf("%02x:%02x:%02x:%02x:%02x:%02x\n",
+	       buf[0], buf[1], buf[2], buf[3], buf[4], buf[5]);
+}
+
+static int maccalc_do_add(int argc, const char *argv[])
+{
+	unsigned char mac[MAC_ADDRESS_LEN];
+	uint32_t t;
+	int err;
+	int i;
+
+	if (argc != 2) {
+		usage();
+		return ERR_INVALID;
+	}
+
+	err = parse_mac(argv[0], mac);
+	if (err)
+		return err;
+
+	i = atoi(argv[1]);
+
+	t = (mac[3] << 16) | (mac[4] << 8) | mac[5];
+	t += i;
```

Comment:
Shouldn't we care about overflow?

Reply:
The code probably assumes that an overflow into the OUI is not a use case.

Reply:
I agree, that the upstream code has room for improvement, but for now, I'd re-use it as-is. And I agree that an overflow into the OUI should not happen.